04cpp3/sec02Templates.md

```diff
@@ -0,0 +1,351 @@
+
+We have to be a little careful when working with this model in the case of templates. We'll explore this using a function template, although the considerations are the same for a class template as well. Let's declare a function in a header file, and write two source files: one which implements the function in the header, and one which makes use of this function. 
+
+First the declaration in `declaration.hpp`:
+```cpp
+#ifndef DECLARATION_HPP 
+#define DECLARATION_HPP
+
+namespace utilFunctions
+{
+    int add(int, int);
+}
+
+#endif
+```
+- It's usually a good idea to put functions and variables which are defined in a header but not part of a class inside a namespace, to avoid potential name clashes. 
+- The `#ifndef`, `#define` and `#endif` are pre-processor directives. This pattern is called an "include guard": it means that the file's contents will be ignored if it has already been included somewhere else in the same compilation unit, so that the contents are not declared twice (which would cause an error). 
+
+Then the implementation in `implementation.hpp`:
+```cpp
+#include "declaration.hpp"
+
+int utilFunctions::add(int a, int b)
+{
+    int c = a + b;
+    return c;
+}
+```
+- We need to include `declaration.hpp` in order to have access to the namespace and function declaration. 
+
+Finally we have another file which will want to use this function, which we'll call `usage.cpp`. 
+```cpp
+#include "declaration.hpp"
+#include <iostream>
+
+int main()
+{
+    int x = 15;
+    int y = 27;
+
+    std::cout << utilFunctions::add(x, y) << std::endl;
+
+    return 0;
+}
+```
+- This also needs to include `declaration.hpp` in order to access the function declaration. 
+- Note that it does **not** need the function implementation: it only needs to know the name of the function and what types it accepts and returns to be able to compile.
+
+We can compile `implementation.cpp` and `usage.cpp` down to two separate object files and link them to produce a fully functional executable. 
+
+This simple addition function could be made much more flexible by also operating on other numeric types, like `float` or `double` for which the addition operator is defined. In this case, rather than writing three separate functions out, all with essentially identical code, we could use a function template. 
+
+Let's update our declaration to use a template:
+```cpp
+#ifndef DECLARATION_HPP 
+#define DECLARATION_HPP
+
+namespace utilFunctions
+{
+    template<typename T>
+    T add(T, T);
+}
+
+#endif
+```
+
+and our implementation:
+```cpp
+#include "declaration.hpp"
+
+template<typename T>
+T utilFunctions::add(T a, T b)
+{
+    T c = a + b;
+    return c;
+}
+```
+
+If we try now to compile and link our executable we will find an error like this:
+```
+undefined reference to `int utilFunctions::add<int>(int, int)'
```

Comment:
Could use the entire error outputted by Cmake to help with the issue of students not knowing how to read Cmake error output?